hail/src/main/scala/is/hail/expr/ir/Binds.scala

```diff
@@ -3,321 +3,411 @@ package is.hail.expr.ir
-      else if (i == 2) FastSeq(accumName -> zero.typ, otherAccumName -> zero.typ)
-      else empty
-    case NDArrayMap(nd, name, _) =>
-      if (i == 1) Array(name -> tcoerce[TNDArray](nd.typ).elementType) else empty
-    case NDArrayMap2(l, r, lName, rName, _, _) => if (i == 2) Array(
-        lName -> tcoerce[TNDArray](l.typ).elementType,
-        rName -> tcoerce[TNDArray](r.typ).elementType,
-      )
-      else empty
-    case CollectDistributedArray(contexts, globals, cname, gname, _, _, _, _) => if (i == 2)
-        Array(cname -> tcoerce[TStream](contexts.typ).elementType, gname -> globals.typ)
-      else empty
-    case TableAggregate(child, _) => if (i == 1) child.typ.globalEnv.m else empty
-    case MatrixAggregate(child, _) => if (i == 1) child.typ.globalEnv.m else empty
-    case TableFilter(child, _) => if (i == 1) child.typ.rowEnv.m else empty
-    case TableGen(contexts, globals, cname, gname, _, _, _) =>
-      if (i == 2) Array(cname -> elementType(contexts.typ), gname -> globals.typ)
-      else empty
-    case TableMapGlobals(child, _) => if (i == 1) child.typ.globalEnv.m else empty
-    case TableMapRows(child, _) => if (i == 1) child.typ.rowEnv.m else empty
-    case TableAggregateByKey(child, _) => if (i == 1) child.typ.globalEnv.m else empty
-    case TableKeyByAndAggregate(child, _, _, _, _) =>
-      if (i == 1) child.typ.globalEnv.m else if (i == 2) child.typ.rowEnv.m else empty
-    case TableMapPartitions(child, g, p, _, _, _) =>
-      if (i == 1) Array(g -> child.typ.globalType, p -> TStream(child.typ.rowType)) else empty
-    case MatrixMapRows(child, _) => if (i == 1) child.typ.rowEnv.bind("n_cols", TInt32).m else empty
-    case MatrixFilterRows(child, _) => if (i == 1) child.typ.rowEnv.m else empty
-    case MatrixMapCols(child, _, _) =>
-      if (i == 1) child.typ.colEnv.bind("n_rows", TInt64).m else empty
-    case MatrixFilterCols(child, _) => if (i == 1) child.typ.colEnv.m else empty
-    case MatrixMapEntries(child, _) => if (i == 1) child.typ.entryEnv.m else empty
-    case MatrixFilterEntries(child, _) => if (i == 1) child.typ.entryEnv.m else empty
-    case MatrixMapGlobals(child, _) => if (i == 1) child.typ.globalEnv.m else empty
-    case MatrixAggregateColsByKey(child, _, _) =>
-      if (i == 1) child.typ.rowEnv.m else if (i == 2) child.typ.globalEnv.m else empty
-    case MatrixAggregateRowsByKey(child, _, _) =>
-      if (i == 1) child.typ.colEnv.m else if (i == 2) child.typ.globalEnv.m else empty
-    case BlockMatrixMap(_, eltName, _, _) => if (i == 1) Array(eltName -> TFloat64) else empty
-    case BlockMatrixMap2(_, _, lName, rName, _, _) =>
-      if (i == 2) Array(lName -> TFloat64, rName -> TFloat64) else empty
-    case _ => empty
-  }
+case class SegregatedBindingEnv[A, B](
+  childEnvWithoutBindings: BindingEnv[A],
+  newBindings: BindingEnv[B],
+) extends GenericBindingEnv[SegregatedBindingEnv[A, B], B] {
+  def unified(implicit ev: BindingEnv[B] <:< BindingEnv[A]): BindingEnv[A] =
```

Comment:
I don't think this constraint it correct.
If `B <:< A`, then it's not true that `BindingEnv[B] <:< BindingEnv[A]`.

Reply:
Revert to type equality - there're too many changes needed to `Env` to make this work properly.

Reply:
> I don't think this constraint it correct. If `B <:< A`, then it's not true that `BindingEnv[B] <:< BindingEnv[A]`.

But I only require `BindingEnv[B] <:< BindingEnv[A]` directly. This is correct, if maybe never more general than type equality in actuality. I'll go back.

Reply:
Which is kinda meaningless unless you intend to subclass BindingEnv